packages/block-editor/src/components/block-list/block-wrapper.js

```diff
@@ -111,6 +111,20 @@ export function useBlockWrapperProps( props = {}, { __unstableIsHtml } = {} ) {
 		}
 	}, [ isSelected, isFirstMultiSelected, isLastMultiSelected ] );
 
+	// Set new block node if it changes.
+	// This effect should happen on every render, so no dependencies should be
+	// added.
+	useEffect( () => {
+		const node = ref.current;
+		setBlockNodes( ( nodes ) => {
+			if ( ! nodes[ clientId ] || nodes[ clientId ] === node ) {
+				return nodes;
+			}
+
+			return { ...nodes, [ clientId ]: node };
+		} );
+	} );
```

Comment:
Instead of checking on every hook call, this seems like the perfect use case for the "ref as function"?

Reply:
We cannot use a function because down the line we're relying on .current, but we could try a setter here.

Reply:
Maybe let's merge this at least so the issue is fixed and follow up.

Reply:
We should also change this to keep all block nodes, not just the selected ones.